proposed/activity-improvements.md

```diff
@@ -0,0 +1,393 @@
+    }
+
+    ...
+
+    if (activity != null)
+    {
+        activityListener.StopActivity(activity);
+    }
+}
+```
+
+Note in the pattern, the code trying to avoid creating the Activity object if there are no listeners interested to listen to such Activity creation. Also, the code later has to check if the activity object is created and then stop it.
+We are simplifying the pattern to something like:
+
+```C#
+    static ActivitySource source =  new ActivitySource("Azure.Core.Http", new Version(4, 0, 0, 0));
+    ...
+
+    // StartActivity can return null if no listener is enabled. Activity now is Disposable too which automatically stop the activity if it is created and started.
+    using (Activity? activity = source.StartActivity("Request")
+    {
+        activity?.AddTag..
+    }
+```
+
+- Last, we are introducing ActivitySource type which can be used to efficiently listening to Activities created from specific component/sources or all sources. It is now reasonable for a listener such as OpenTelemetry to observe Activity instrumentation from any code component in the app by the `ActivitySource.AddActivityListener()`. Previously OpenTelemetry would have needed appropriate  knowledge of the strings that would be used for DiagnosticSource name and IsEnabled() callbacks for each Activity it wanted to receive callbacks for. When trying to create a new Activity object through the new introduced factory methods `ActivitySource.StartActivity`, the listener will have the opportunity to get a callback to decide if interested in having the Activity object get created or it can be sampled out to avoid the object creation if nobody else is listening or other listeners are not interested in such activity too. We provide a way to listen to ActivitySource object creation too to get the opportunity for listening to Activities created for such sources. Also, we are providing `Activity.Source` property to return the source of the Activity. Activities created using `new Activity(...)` will be attached to a singleton source object which can be used too to listen to the Activities created using the Activity constructors.
+ActivitySource will provide a way to listen to other sources creation using `ActivitySource.AddListener` which will notify the listeners for every ActivitySource object previously created or objects are going to be created in the future.
+
+Here is example of how to create a listener for ActivitySource creation event.
+
+```C#
+    ActivitySource.AddListener(source => { ... });
+```
+
+Here is example of how to create a listener for Activities events.
+
+```C#
+    static ActivitySource source =  new ActivitySource("MyLibrary.Subcomponent", new Version(1, 0, 0, 0));
+
+    ...
+
+    IDisposable listener = source.AddActivityListener(
+                            getActivityCreationOption: (source, name, kind, parentContext, tags, links) => ActivityCreationOptions.AllowData,
```

Comment:
Sorry, I don't remember that. are you suggesting to have extra callback taking the full parameters like? what scenario exactly this will get used with?

```C#
getActivityCreationOption: (source, name, kind, parent, tags, links) // parent is string
```

Reply:
done